sv.c

```diff
@@ -4771,6 +4776,272 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
+                /* Source was SVf_NOK|SVp_NOK|SVp_POK but not SVf_POK, meaning
+                   a value set as floating point and later stringified, where
+                  the value happens to be one of the few that we know aren't
+                  affected by the numeric locale, hence we can cache the
+                  stringification. Currently that's  +Inf, -Inf and NaN, but
+                  conceivably we might extend this to -9 .. +9 (excluding -0).
+                  So mark destination the same: */
+                SvFLAGS(dsv) &= ~SVf_POK;
+            }
+            break;
+        case SVp_POK|SVp_IOK|SVp_NOK:
+            assert(SvTYPE(ssv) != SVt_INVLIST);
+
+            S_sv_freshcopy_POK(aTHX_ dsv, ssv, flags);
+        {
+            const char *vstr_pv;
+            STRLEN vstr_len;
+            if ((vstr_pv = SvVSTRING(ssv, vstr_len))) {
+                sv_magic(dsv, NULL, PERL_MAGIC_vstring, vstr_pv, vstr_len);
+                SvRMAGICAL_on(dsv);
+            }
+        }
+            SvIV_set(dsv, SvIVX(ssv));
+            SvNV_set(dsv, SvNVX(ssv));
+
+            SvFLAGS(dsv) |= (sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV|SVf_NOK|SVp_NOK));
+            if ((sflags & SVf_IOK) && !(sflags & SVf_POK)) {
+                /* Source was SVf_IOK|SVp_IOK|SVp_POK but not SVf_POK, meaning
+                   a value set as an integer and later stringified. So mark
+                   destination the same: */
+                SvFLAGS(dsv) &= ~SVf_POK;
+            } else if ((sflags & SVf_NOK) && !(sflags & SVf_POK)) {
+                /* Source was SVf_NOK|SVp_NOK|SVp_POK but not SVf_POK, meaning
+                   a value set as floating point and later stringified, where
+                  the value happens to be one of the few that we know aren't
+                  affected by the numeric locale, hence we can cache the
+                  stringification. Currently that's  +Inf, -Inf and NaN, but
+                  conceivably we might extend this to -9 .. +9 (excluding -0).
+                  So mark destination the same: */
+                SvFLAGS(dsv) &= ~SVf_POK;
+            }
+            break;
+        default:
+            if(!SvOK(ssv))
+                break;
+            /* Some cases seem so rare that we may as well let
+             * sv_setsv_flags deal with them. For example:
+             *     SVp_IOK|SVp_NOK
+             *
+             * Some cases are (currently) not naturally occurring:
+             *     SVp_POK|SVppv_STATIC
+             *     SVp_POK|SVp_IOK|SVppv_STATIC
+             *     SVp_POK|SVp_NOK|SVppv_STATIC
+             *
+             * Other cases are also rare but also trickier to handle,
+             * so keeps this function smaller to not even try. */
+            sv_setsv_flags(dsv,ssv,flags);
+            return dsv;
+        case SVp_NOK:
+            SvNV_set(dsv, SvNVX(ssv));
+            SvFLAGS(dsv) |= (sflags & (SVf_NOK|SVp_NOK));
```

Comment:
why read and extract bits from `sflags`? does SVp_NOK WITHOUT SVf_NOK and no other "OK" bits, does that have any meaning on a PP level? is it ever encountered at runtime in a blead/stable perl running normal production PP and CPAN XS code? do we need to consider buggy/weird/beginner CPAN XS code that would make such a SV* head or not?

`|=` with a literal on the right side, is much more efficient that bit extraction from var sflags, in CC optimized C code.

Reply:
Basically it's defensive coding because I don't know what flags are set by XS code out in the wild. Or to what extend we should support such code.

Happy for those discussions to happen, I just didn't want to arbitrarily decide in private. :smile: